File: examples/FsharpExamples/Telemetry.fs
module Telemetry

open System
open System.Threading.Tasks
open OpenTelemetry
open Pulsar.Client.Api

open Pulsar.Client.Common
open System.Text
open Pulsar.Client.Otel.OTelProducerInterceptor
open Pulsar.Client.Otel.OtelConsumerInterceptor
open OpenTelemetry.Resources
open OpenTelemetry.Trace

let runTelemetry()=
    let serviceUrl = "pulsar://localhost:6650"
    let subscriptionName = "my-subscription"
    let topicName = $"my-topic-{DateTime.Now.Ticks}"
    let producerSourceName = "pulsar.producer"
    let consumerSourceName = "pulsar.consumer"
    
    Sdk.CreateTracerProviderBuilder()
                .AddSource(producerSourceName, consumerSourceName)
                .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("telemetry"))
                .AddConsoleExporter().Build() |> ignore
    task {
        let! client =
            PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync()
        use prodIntercept = new OTelProducerInterceptor<byte[]>(producerSourceName, PulsarClient.Logger)
        use consIntercept = new OTelConsumerInterceptor<byte[]>(consumerSourceName, PulsarClient.Logger)

        let! producer =
            client.NewProducer()
                .Intercept(prodIntercept)
                .Topic(topicName)
                .CreateAsync()

        let! consumer =
            client.NewConsumer()
                .Topic(topicName)
                .Intercept(consIntercept)
                .SubscriptionName(subscriptionName)
                .SubscriptionType(SubscriptionType.Exclusive)
                .SubscribeAsync()
        let! _ =
            [|
                producer.SendAsync(Encoding.UTF8.GetBytes("Sent 1 from F# at " + DateTime.Now.ToString()))
                producer.SendAsync(Encoding.UTF8.GetBytes("Sent 2 from F# at " + DateTime.Now.ToString()))
            |] |> Task.WhenAll
       
        let! msgs =
            [|
                consumer.ReceiveAsync()
                consumer.ReceiveAsync()
            |] |> Task.WhenAll
            
        for message in msgs do
            message.Data |> Encoding.UTF8.GetString |> Console.WriteLine
        do! consumer.AcknowledgeCumulativeAsync(msgs.[1].MessageId)
    }

File: examples/CsharpExamples/Telemetry.cs
using System;
using System.Text;
using System.Threading.Tasks;
using OpenTelemetry;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;
using Pulsar.Client.Api;
using Pulsar.Client.Otel;

namespace CsharpExamples
{
    internal class Telemetry
    {
        internal static async Task RunTelemetry()
        {
            const string serviceUrl = "pulsar://my-pulsar-cluster:32268";
            const string subscriptionName = "my-subscription";
            const string producerSourceName = "pulsar.producer";
            const string consumerSourceName = "pulsar.consumer";
            
            var topicName = $"my-topic-{DateTime.Now.Ticks}";

            Sdk.CreateTracerProviderBuilder()
                .AddSource(producerSourceName, consumerSourceName)
                .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("telemetry"))
                .AddConsoleExporter().Build();
            
            using var prodIntercept = 
                new OTelProducerInterceptor.OTelProducerInterceptor<byte[]>(producerSourceName, PulsarClient.Logger);
            using var consIntercept =
                new OtelConsumerInterceptor.OTelConsumerInterceptor<byte[]>(consumerSourceName, PulsarClient.Logger);

            var client = await new PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync();

            var producer = await client.NewProducer()
                .Intercept(prodIntercept)
                .Topic(topicName)
                .CreateAsync();

            var consumer = await client.NewConsumer()
                .Intercept(consIntercept)
                .Topic(topicName)
                .SubscriptionName(subscriptionName)
                .SubscribeAsync();

            var msgIdss = await Task.WhenAll(new[]
            {
                producer.SendAsync(Encoding.UTF8.GetBytes($"Sent 1 from C# at '{DateTime.Now}'")),
                producer.SendAsync(Encoding.UTF8.GetBytes($"Sent 2 from C# at '{DateTime.Now}'"))
            });
            
            var msgs = await Task.WhenAll(new[]
            {
                consumer.ReceiveAsync(),
                consumer.ReceiveAsync()
            });

            foreach (var msg in msgs)
            {
                Console.WriteLine(Encoding.UTF8.GetString(msg.GetValue()));
            }

            await consumer.AcknowledgeCumulativeAsync(msgs[1].MessageId);
        }
    }
}